### ISSUE TITLE
Optionally download source code along with toolchain

### ISSUE THREAD

**Issue by Grawp** [rust-lang-nursery/rustup.rs#37]
Rust sources comes handy when you are using **racer** or when you are cross-compiling etc...
It would be nice to have a possibility to download/upgrade rust sources along with the toolchain.

**Comment by matklad** [comment]
Maybe it's a good idea even to download them by default. `libstd` should not be that big, and if you are using multirust then you are most likely a developer and need sources for completion anyway.

**Comment by steveklabnik** [comment]
> if you are using multirust then you are most likely a developer and need sources for completion anyway. 

Not all developers use completion.

**Comment by matklad** [comment]
>Not all developers use completion.

Hm, I think that overwhelming majority uses at least hippie-expand or some alternative in their editor of choice. 

Not all developers setup semantic completion, but I think it is because hippie-expand works out of the box, and almost anything else requires some awkward setup to actually be useful. So I think your argument works backwards: if we download sources by default, completion setup becomes easier and more developers use it :)

However I think that this may be not the best option here: see https://internals.rust-lang.org/t/what-do-you-think-about-publishing-libstd-on-crates-io/3227

**Comment by steveklabnik** [comment]
> Hm, I think that overwhelming majority uses at least hippie-expand

I've never even heard of hippie-expand. :smile: I think it really depends on the circles you run in.

**Comment by matklad** [comment]
>I've never even heard of hippie-expand. 
>or some alternative in their editor of choice.

In vim it is `Ctrl-N` and called keyword completion I think

**Comment by matklad** [comment]
@brson I'd like to try to implement this unless libstd and related crates are uploaded to crates.io in the nearest future :)

**Comment by toothbrush7777777** [comment]
@matklad Any updates?

**Comment by matklad** [comment]
@toothbrush7777777 no, unfortunately other things have happened :) At the same time the https://github.com/rust-lang/rfcs/pull/1133/ RFC looks like a better solution for the problem.

**Comment by Diggsey** [comment]
Prerequisite: rust-lang/rust-buildbot#102

**Comment by aka-demik** [comment]
rust-lang/rust-buildbot#102 was merged.

**Comment by Diggsey** [comment]
@aka-demik Yep, the next nightly will be the first to contain source packages. Then we can test #659 against those source packages and hopefully get it merged.

I'm confident we'll be able to install source packages with the next release of rustup.

**Comment by Diggsey** [comment]
So the good news is that the rustup support for adding components (eg. source packages) was merged. The bad news is that it uncovered a problem in the new `rust-src` package generated upstream (see rust-lang/rust#35840)  :disappointed:

**Comment by jplatte** [comment]
`rustup component add rust-src` seems to work now! :)

I'll have to install racer via cargo it seems, as `rustup run nightly-x86_64-unknown-linux-gnu racer complete std::io::B` errors with a command not found when racer is in /usr/bin. But that shouldn't be a problem. I'll report back when it's installed and I've played around with it a bit!

**Comment by jplatte** [comment]
Okay, so installing things works... But I expected a more out-of-the-box experience. Right now on arch linux it's actually harder to set up racer with rust sources installed by rustup than setting it up by installing rust-src from the AUR, because with rustup the sources are now in a much more obscure path – `$HOME/.multirust/toolchains/$TOOLCHAIN/lib/rustlib/src/rust/src` instead of just `/usr/src/rust/src/`. I don't know where this would best be fixed, as apparently `rustup run` isn't for what I thought it was so exporting it from there probably wouldn't make sense. But I think it's important that it is fixed somewhere.

I'll leave this here, in case someone on a different distribution wants to use the sources installed by rustup. It will set `RUST_SRC_PATH` to where rustup installed them (assuming you installed them for your default toolchain) when put in your shells config file (usually `~/.bashrc`; requires re-login to take effect):

```
RUST_DEFAULT_TOOLCHAIN=$(rustup show | grep '(default)' | head -1 | awk '{ print $1 }')
export RUST_SRC_PATH="$HOME/.multirust/toolchains/$RUST_DEFAULT_TOOLCHAIN/lib/rustlib/src/rust/src"
```

**Comment by Diggsey** [comment]
\o/ finally!

@jplatte I've opened this issue: phildawes/racer#595 which will make this zero setup.

**Comment by ariasuni** [comment]
Well, this doesn’t work very well for the Atom integration, which use either default `/usr/local/src/rust/src` or some static string of text (so I can’t use the `--print sysroot` trick).

**Comment by WilsonGiese** [comment]
@ariasuni I got around this by creating a symlink. Added a little script in my bash profile to keep it updated as well: 
```
if [ "$(readlink -- "/usr/local/src/rust/src")" = "" ]; then
    $(ln -s $RUST_SRC_PATH /usr/local/src/rust)
elif [ "$(readlink -- "/usr/local/src/rust/src")" != "$RUST_SRC_PATH" ]; then
    rm /usr/local/src/rust/src
    $(ln -s $RUST_SRC_PATH /usr/local/src/rust)
fi
```

Silly, but it works (:
